## Python Data Types and Structure

Data types are the classification or categorization of data items. It represents the kind of value that tells what operations can be performed on a particular data. Since everything is an object in Python programming, data types are actually classes and variables are instance (object) of these classes.

Following are the standard or built-in data type of Python:

* Numeric
* String
* List
* Dictionary
* Tuples
* Set
* Boolean

# Numbers and more in Python!

In this lecture, we will learn about numbers in Python and how to use them.

We'll learn about the following topics:

    1.) Types of Numbers in Python
    2.) Basic Arithmetic
    3.) Differences between classic division and floor division
    4.) Object Assignment in Python

## Types of numbers

Python has various "types" of numbers (numeric literals). We'll mainly focus on integers and floating point numbers.

Integers are just whole numbers, positive or negative. For example: 2 and -2 are examples of integers.

Floating point numbers in Python are notable because they have a decimal point in them, or use an exponential (e) to define the number. For example 2.0 and -2.1 are examples of floating point numbers. 4E2 (4 times 10 to the power of 2) is also an example of a floating point number in Python.

Throughout this course we will be mainly working with integers or simple float number types.

Here is a table of the two main types we will spend most of our time working with some examples:

<table>
<tr>
    <th>Examples</th> 
    <th>Number "Type"</th>
</tr>

<tr>
    <td>1,2,-5,1000</td>
    <td>Integers</td> 
</tr>

<tr>
    <td>1.2,-0.5,2e2,3E2</td> 
    <td>Floating-point numbers</td> 
</tr>
 </table>

 
 
Now let's start with some basic arithmetic.

### Basic Arithmetic

In [2]:
print('Hello something world')

Hello something world


In [1]:
# Subtraction
2-20

-18

In [3]:
# Multiplication
2*2

4

In [4]:
# Division
3/2

1.5

In [3]:
# Floor Division
7/4

1.75

**Whoa! What just happened? Last time I checked, 7 divided by 4 equals 1.75 not 1!**

The reason we get this result is because we are using "*floor*" division. The // operator (two forward slashes) truncates the decimal without rounding, and returns an integer result.

**So what if we just want the remainder after division?**

In [3]:
# Modulo
7%4.5

2.5

4 goes into 7 once, with a remainder of 3. The % operator returns the remainder after division.

### Arithmetic continued

In [7]:
# Powers
2**3

8

In [8]:
# Can also do roots this way
4**0.5

2.0

In [9]:
# Order of Operations followed in Python
2 + 10 * 10 + 3

105

In [10]:
# Can use parentheses to specify orders
(2+10) * (10+3)

156

## Variable Assignments

Now that we've seen how to use numbers in Python as a calculator let's see how we can assign names and create variables.

We use a single equals sign to assign labels to variables. Let's see a few examples of how we can do this.

In [11]:
# Let's create an object called "a" and assign it the number 5
a = 5

Now if I call *a* in my Python script, Python will treat it as the number 5.

In [12]:
# Adding the objects
a+a

10

What happens on reassignment? Will Python let us write it over?

In [13]:
# Reassignment
a = 10

In [14]:
# Check
a

10

Yes! Python allows you to write over assigned variable names. We can also use the variables themselves when doing the reassignment. Here is an example of what I mean:

In [15]:
# Check
a

10

In [16]:
# Use A to redefine A
a = a + a

In [17]:
# Check 
a

20

The names you use when creating these labels need to follow a few rules:

    1. Names can not start with a number.
    2. There can be no spaces in the name, use _ instead.
    3. Can't use any of these symbols :'",<>/?|\()!@#$%^&*~-+
    4. It's considered best practice (PEP8) that names are lowercase.
    5. Avoid using the characters 'l' (lowercase letter el), 'O' (uppercase letter oh), 
       or 'I' (uppercase letter eye) as single character variable names.
    6. Avoid using words that have special meaning in Python like "list" and "str"


Using variable names can be a very useful way to keep track of different variables in Python. For example:

In [18]:
# Use object names to keep better track of what's going on in your code!
my_income = 100

tax_rate = 0.1

my_taxes = my_income*tax_rate

In [19]:
# Show my taxes!
my_taxes

10.0

In [ ]:
a = 3
print(a)
type(a)

In [ ]:
b = 4.56
print(b)
type(b)

In [ ]:
print(type(a))
print(type(b))

***Try dir(a) and dir(b) . This command lists the functions available for these objects.***

In [ ]:
dir(a)

So what have we learned? We learned some of the basics of numbers in Python. We also learned how to do arithmetic and use Python as a basic calculator. We then wrapped it up with learning about Variable Assignment in Python.

Up next we'll learn about Strings!

# Strings

Strings are used in Python to record text information, such as names. Strings in Python are actually a *sequence*, which basically means Python keeps track of every element in the string as a sequence. For example, Python understands the string "hello' to be a sequence of letters in a specific order. This means we will be able to use indexing to grab particular letters (like the first letter, or the last letter).

This idea of a sequence is an important one in Python and we will touch upon it later on in the future.

In this lecture we'll learn about the following:

    1.) Creating Strings
    2.) Printing Strings
    3.) String Indexing and Slicing
    4.) String Properties
    5.) String Methods
    6.) Print Formatting

## Creating a String
To create a string in Python you need to use either single quotes or double quotes. For example:

In [1]:
# Single word
'hello'

'hello'

In [2]:
# Entire phrase 
'This is also a string'

'This is also a string'

In [3]:
# We can also use double quote
"String built with double quotes"

'String built with double quotes'

In [4]:
# Be careful with quotes!
' I'm using single quotes, but this will create an error'

SyntaxError: invalid syntax (<ipython-input-4-da9a34b3dc31>, line 2)

The reason for the error above is because the single quote in <code>I'm</code> stopped the string. You can use combinations of double and single quotes to get the complete statement.

In [5]:
"Now I'm ready to use the single quotes inside a string!"

"Now I'm ready to use the single quotes inside a string!"

Now let's learn about printing strings!

## Printing a String

Using Jupyter notebook with just a string in a cell will automatically output strings, but the correct way to display strings in your output is by using a print function.

In [6]:
# We can simply declare a string
'Hello World'

'Hello World'

In [7]:
# Note that we can't output multiple strings this way
'Hello World 1'
'Hello World 2'

'Hello World 2'

We can use a print statement to print a string.

In [8]:
print('Hello World 1')
print('Hello World 2')
print('Use \n to print a new line')
print('\n')
print('See what I mean?')

Hello World 1
Hello World 2
Use 
 to print a new line


See what I mean?


## String Basics

We can also use a function called len() to check the length of a string!

In [9]:
len('Hello World')

11

Python's built-in len() function counts all of the characters in the string, including spaces and punctuation.

## String Indexing
We know strings are a sequence, which means Python can use indexes to call parts of the sequence. Let's learn how this works.

In Python, we use brackets <code>[]</code> after an object to call its index. We should also note that indexing starts at 0 for Python. Let's create a new object called <code>s</code> and then walk through a few examples of indexing.

In [10]:
# Assign s as a string
s = 'Hello World'

In [11]:
#Check
s

'Hello World'

In [12]:
# Print the object
print(s) 

Hello World


Let's start indexing!

In [13]:
# Show first element (in this case a letter)
s[0]

'H'

In [14]:
s[1]

'e'

In [15]:
s[2]

'l'

We can use a <code>:</code> to perform *slicing* which grabs everything up to a designated point. For example:

In [16]:
# Grab everything past the first term all the way to the length of s which is len(s)
s[1:]

'ello World'

In [17]:
# Note that there is no change to the original s
s

'Hello World'

In [18]:
# Grab everything UP TO the 3rd index
s[:3]

'Hel'

Note the above slicing. Here we're telling Python to grab everything from 0 up to 3. It doesn't include the 3rd index. You'll notice this a lot in Python, where statements and are usually in the context of "up to, but not including".

In [19]:
#Everything
s[:]

'Hello World'

We can also use negative indexing to go backwards.

In [20]:
# Last letter (one index behind 0 so it loops back around)
s[-1]

'd'

In [21]:
# Grab everything but the last letter
s[:-1]

'Hello Worl'

We can also use index and slice notation to grab elements of a sequence by a specified step size (the default is 1). For instance we can use two colons in a row and then a number specifying the frequency to grab elements. For example:

In [22]:
# Grab everything, but go in steps size of 1
s[::1]

'Hello World'

In [23]:
# Grab everything, but go in step sizes of 2
s[::2]

'HloWrd'

In [24]:
# We can use this to print a string backwards
s[::-1]

'dlroW olleH'

## String Properties
It's important to note that strings have an important property known as *immutability*. This means that once a string is created, the elements within it can not be changed or replaced. For example:

In [25]:
s

'Hello World'

In [26]:
# Let's try to change the first letter to 'x'
s[0] = 'x'

TypeError: 'str' object does not support item assignment

Notice how the error tells us directly what we can't do, change the item assignment!

Something we *can* do is concatenate strings!

In [27]:
s

'Hello World'

In [28]:
# Concatenate strings!
s + ' concatenate me!'

'Hello World concatenate me!'

In [29]:
# We can reassign s completely though!
s = s + ' concatenate me!'

In [30]:
print(s)

Hello World concatenate me!


In [31]:
s

'Hello World concatenate me!'

We can use the multiplication symbol to create repetition!

In [32]:
letter = 'z'

In [33]:
letter*10

'zzzzzzzzzz'

## Basic Built-in String methods

Objects in Python usually have built-in methods. These methods are functions inside the object (we will learn about these in much more depth later) that can perform actions or commands on the object itself.

We call methods with a period and then the method name. Methods are in the form:

object.method(parameters)

Where parameters are extra arguments we can pass into the method. Don't worry if the details don't make 100% sense right now. Later on we will be creating our own objects and functions!

Here are some examples of built-in methods in strings:

In [34]:
s

'Hello World concatenate me!'

In [35]:
# Upper Case a string
s.upper()

'HELLO WORLD CONCATENATE ME!'

In [36]:
# Lower case
s.lower()

'hello world concatenate me!'

In [37]:
# Split a string by blank space (this is the default)
s.split()

['Hello', 'World', 'concatenate', 'me!']

In [38]:
# Split by a specific element (doesn't include the element that was split on)
s.split('W')

['Hello ', 'orld concatenate me!']

There are many more methods than the ones covered here. Visit the Advanced String section to find out more!

## Print Formatting

We can use the .format() method to add formatted objects to printed string statements. 

The easiest way to show this is through an example:

In [39]:
'Insert another string with curly brackets: {}'.format('The inserted string')

'Insert another string with curly brackets: The inserted string'

We will revisit this string formatting topic in later sections when we are building our projects!

## Next up: Lists!

# Lists

Earlier when discussing strings we introduced the concept of a *sequence* in Python. Lists can be thought of the most general version of a *sequence* in Python. Unlike strings, they are mutable, meaning the elements inside a list can be changed!

In this section we will learn about:
    
    1.) Creating lists
    2.) Indexing and Slicing Lists
    3.) Basic List Methods
    4.) Nesting Lists
    5.) Introduction to List Comprehensions

Lists are constructed with brackets [] and commas separating every element in the list.

Let's go ahead and see how we can construct lists!

In [1]:
# Assign a list to an variable named my_list
my_list = [1,2,3]

We just created a list of integers, but lists can actually hold different object types. For example:

In [2]:
my_list = ['A string',23,100.232,'o']

Just like strings, the len() function will tell you how many items are in the sequence of the list.

In [3]:
len(my_list)

4

### Indexing and Slicing
Indexing and slicing work just like in strings. Let's make a new list to remind ourselves of how this works:

In [4]:
my_list = ['one','two','three',4,5]

In [5]:
# Grab element at index 0
my_list[0]

'one'

In [6]:
# Grab index 1 and everything past it
my_list[1:]

['two', 'three', 4, 5]

In [7]:
# Grab everything UP TO index 3
my_list[:3]

['one', 'two', 'three']

We can also use + to concatenate lists, just like we did for strings.

In [8]:
my_list + ['new item']

['one', 'two', 'three', 4, 5, 'new item']

Note: This doesn't actually change the original list!

In [9]:
my_list

['one', 'two', 'three', 4, 5]

You would have to reassign the list to make the change permanent.

In [10]:
# Reassign
my_list = my_list + ['add new item permanently']

In [11]:
my_list

['one', 'two', 'three', 4, 5, 'add new item permanently']

We can also use the * for a duplication method similar to strings:

In [12]:
# Make the list double
my_list * 2

['one',
 'two',
 'three',
 4,
 5,
 'add new item permanently',
 'one',
 'two',
 'three',
 4,
 5,
 'add new item permanently']

In [13]:
# Again doubling not permanent
my_list

['one', 'two', 'three', 4, 5, 'add new item permanently']

## Basic List Methods

If you are familiar with another programming language, you might start to draw parallels between arrays in another language and lists in Python. Lists in Python however, tend to be more flexible than arrays in other languages for a two good reasons: they have no fixed size (meaning we don't have to specify how big a list will be), and they have no fixed type constraint (like we've seen above).

Let's go ahead and explore some more special methods for lists:

In [14]:
# Create a new list
list1 = [1,2,3]

Use the **append** method to permanently add an item to the end of a list:

In [15]:
# Append
list1.append('append me!')

In [16]:
# Show
list1

[1, 2, 3, 'append me!']

Use **pop** to "pop off" an item from the list. By default pop takes off the last index, but you can also specify which index to pop off. Let's see an example:

In [17]:
# Pop off the 0 indexed item
list1.pop(0)

1

In [18]:
# Show
list1

[2, 3, 'append me!']

In [19]:
# Assign the popped element, remember default popped index is -1
popped_item = list1.pop()

In [20]:
popped_item

'append me!'

In [21]:
# Show remaining list
list1

[2, 3]

It should also be noted that lists indexing will return an error if there is no element at that index. For example:

In [22]:
list1[100]

IndexError: list index out of range

We can use the **sort** method and the **reverse** methods to also effect your lists:

In [23]:
new_list = ['a','e','x','b','c']

In [24]:
#Show
new_list

['a', 'e', 'x', 'b', 'c']

In [25]:
# Use reverse to reverse order (this is permanent!)
new_list.reverse()

In [26]:
new_list

['c', 'b', 'x', 'e', 'a']

In [27]:
# Use sort to sort the list (in this case alphabetical order, but for numbers it will go ascending)
new_list.sort()

In [28]:
new_list

['a', 'b', 'c', 'e', 'x']

## Nesting Lists
A great feature of of Python data structures is that they support *nesting*. This means we can have data structures within data structures. For example: A list inside a list.

Let's see how this works!

In [29]:
# Let's make three lists
lst_1=[1,2,3]
lst_2=[4,5,6]
lst_3=[7,8,9]

# Make a list of lists to form a matrix
matrix = [lst_1,lst_2,lst_3]

In [30]:
# Show
matrix

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]

We can again use indexing to grab elements, but now there are two levels for the index. The items in the matrix object, and then the items inside that list!

In [31]:
# Grab first item in matrix object
matrix[0]

[1, 2, 3]

In [32]:
# Grab first item of the first item in the matrix object
matrix[0][0]

1

# List Comprehensions
Python has an advanced feature called list comprehensions. They allow for quick construction of lists. To fully understand list comprehensions we need to understand for loops. So don't worry if you don't completely understand this section, and feel free to just skip it since we will return to this topic later.

But in case you want to know now, here are a few examples!

In [33]:
# Build a list comprehension by deconstructing a for loop within a []
first_col = [row[0] for row in matrix]

In [34]:
first_col

[1, 4, 7]

We used a list comprehension here to grab the first element of every row in the matrix object. We will cover this in much more detail later on!

For more advanced methods and features of lists in Python, check out the Advanced Lists section later on in this course!

# Dictionaries

We've been learning about *sequences* in Python but now we're going to switch gears and learn about *mappings* in Python. If you're familiar with other languages you can think of these Dictionaries as hash tables. 

This section will serve as a brief introduction to dictionaries and consist of:

    1.) Constructing a Dictionary
    2.) Accessing objects from a dictionary
    3.) Nesting Dictionaries
    4.) Basic Dictionary Methods

So what are mappings? Mappings are a collection of objects that are stored by a *key*, unlike a sequence that stored objects by their relative position. This is an important distinction, since mappings won't retain order since they have objects defined by a key.

A Python dictionary consists of a key and then an associated value. That value can be almost any Python object.


## Constructing a Dictionary
Let's see how we can construct dictionaries to get a better understanding of how they work!

In [1]:
# Make a dictionary with {} and : to signify a key and a value
my_dict = {'key1':'value1','key2':'value2'}

In [2]:
# Call values by their key
my_dict['key2']

'value2'

Its important to note that dictionaries are very flexible in the data types they can hold. For example:

In [3]:
my_dict = {'key1':123,'key2':[12,23,33],'key3':['item0','item1','item2']}

In [4]:
# Let's call items from the dictionary
my_dict['key3']

['item0', 'item1', 'item2']

In [5]:
# Can call an index on that value
my_dict['key3'][0]

'item0'

In [6]:
# Can then even call methods on that value
my_dict['key3'][0].upper()

'ITEM0'

We can affect the values of a key as well. For instance:

In [7]:
my_dict['key1']

123

In [8]:
# Subtract 123 from the value
my_dict['key1'] = my_dict['key1'] - 123

In [9]:
#Check
my_dict['key1']

0

A quick note, Python has a built-in method of doing a self subtraction or addition (or multiplication or division). We could have also used += or -= for the above statement. For example:

In [10]:
# Set the object equal to itself minus 123 
my_dict['key1'] -= 123
my_dict['key1']

-123

We can also create keys by assignment. For instance if we started off with an empty dictionary, we could continually add to it:

In [11]:
# Create a new dictionary
d = {}

In [12]:
# Create a new key through assignment
d['animal'] = 'Dog'

In [13]:
# Can do this with any object
d['answer'] = 42

In [14]:
#Show
d

{'animal': 'Dog', 'answer': 42}

## Nesting with Dictionaries

Hopefully you're starting to see how powerful Python is with its flexibility of nesting objects and calling methods on them. Let's see a dictionary nested inside a dictionary:

In [15]:
# Dictionary nested inside a dictionary nested inside a dictionary
d = {'key1':{'nestkey':{'subnestkey':'value'}}}

Wow! That's a quite the inception of dictionaries! Let's see how we can grab that value:

In [16]:
# Keep calling the keys
d['key1']['nestkey']['subnestkey']

'value'

## A few Dictionary Methods

There are a few methods we can call on a dictionary. Let's get a quick introduction to a few of them:

In [17]:
# Create a typical dictionary
d = {'key1':1,'key2':2,'key3':3}

In [18]:
# Method to return a list of all keys 
d.keys()

dict_keys(['key1', 'key2', 'key3'])

In [19]:
# Method to grab all values
d.values()

dict_values([1, 2, 3])

In [20]:
# Method to return tuples of all items  (we'll learn about tuples soon)
d.items()

dict_items([('key1', 1), ('key2', 2), ('key3', 3)])

Hopefully you now have a good basic understanding how to construct dictionaries. There's a lot more to go into here, but we will revisit dictionaries at later time. After this section all you need to know is how to create a dictionary and how to retrieve values from it.

# Tuples

In Python tuples are very similar to lists, however, unlike lists they are *immutable* meaning they can not be changed. You would use tuples to present things that shouldn't be changed, such as days of the week, or dates on a calendar. 

In this section, we will get a brief overview of the following:

    1.) Constructing Tuples
    2.) Basic Tuple Methods
    3.) Immutability
    4.) When to Use Tuples

You'll have an intuition of how to use tuples based on what you've learned about lists. We can treat them very similarly with the major distinction being that tuples are immutable.

## Constructing Tuples

The construction of a tuples use () with elements separated by commas. For example:

In [1]:
# Create a tuple
t = (1,2,3)

In [2]:
# Check len just like a list
len(t)

3

In [3]:
# Can also mix object types
t = ('one',2)

# Show
t

('one', 2)

In [4]:
# Use indexing just like we did in lists
t[0]

'one'

In [5]:
# Slicing just like a list
t[-1]

2

## Basic Tuple Methods

Tuples have built-in methods, but not as many as lists do. Let's look at two of them:

In [6]:
# Use .index to enter a value and return the index
t.index('one')

0

In [7]:
# Use .count to count the number of times a value appears
t.count('one')

1

## Immutability

It can't be stressed enough that tuples are immutable. To drive that point home:

In [8]:
t[0]= 'change'

TypeError: 'tuple' object does not support item assignment

Because of this immutability, tuples can't grow. Once a tuple is made we can not add to it.

In [9]:
t.append('nope')

AttributeError: 'tuple' object has no attribute 'append'

## When to use Tuples

You may be wondering, "Why bother using tuples when they have fewer available methods?" To be honest, tuples are not used as often as lists in programming, but are used when immutability is necessary. If in your program you are passing around an object and need to make sure it does not get changed, then a tuple becomes your solution. It provides a convenient source of data integrity.

You should now be able to create and use tuples in your programming as well as have an understanding of their immutability.

Up next Sets and Booleans!!

# Set and Booleans

There are two other object types in Python that we should quickly cover: Sets and Booleans. 

## Sets

Sets are an unordered collection of *unique* elements. We can construct them by using the set() function. Let's go ahead and make a set to see how it works

In [1]:
x = set()

In [2]:
# We add to sets with the add() method
x.add(1)

In [3]:
#Show
x

{1}

Note the curly brackets. This does not indicate a dictionary! Although you can draw analogies as a set being a dictionary with only keys.

We know that a set has only unique entries. So what happens when we try to add something that is already in a set?

In [4]:
# Add a different element
x.add(2)

In [5]:
#Show
x

{1, 2}

In [6]:
# Try to add the same element
x.add(1)

In [7]:
#Show
x

{1, 2}

Notice how it won't place another 1 there. That's because a set is only concerned with unique elements! We can cast a list with multiple repeat elements to a set to get the unique elements. For example:

In [8]:
# Create a list with repeats
list1 = [1,1,2,2,3,4,5,6,1,1]

In [9]:
# Cast as set to get unique values
set(list1)

{1, 2, 3, 4, 5, 6}

## Booleans

Python  comes with Booleans (with predefined True and False displays that are basically just the integers 1 and 0). It also has a placeholder object called None. Let's walk through a few quick examples of Booleans (we will dive deeper into them later in this course).

In [10]:
# Set object to be a boolean
a = True

In [11]:
#Show
a

True

We can also use comparison operators to create booleans. We will go over all the comparison operators later on in the course.

In [12]:
# Output is boolean
1 > 2

False

We can use None as a placeholder for an object that we don't want to reassign yet:

In [13]:
# None placeholder
b = None

In [14]:
# Show
print(b)

None


Thats it! You should now have a basic understanding of Python objects and data structure types. Next, go ahead and do the assessment test!